README.md
<div><h1>duckfficer-method-di</h1></div>

<p>
    <a href="https://www.npmjs.com/package/duckfficer-method-di" target="_blank"><img src="https://img.shields.io/npm/v/duckfficer-method-di.svg" alt="Version"></a>
<a href="http://opensource.org/licenses" target="_blank"><img src="http://img.shields.io/badge/License-MIT-brightgreen.svg"></a>
</p>

<p>
    loads duckfficer-method entities from a path to provide a pleasure-di container
</p>

## Installation

```sh
$ npm i duckfficer-method-di --save
# or
$ yarn add duckfficer-method-di
```

## Features

- [builds an object appending all method found in given paths](#builds-an-object-appending-all-method-found-in-given-paths)
- [validates input](#validates-input)
- [validates output](#validates-output)
- [emits events](#emits-events)


<a name="builds-an-object-appending-all-method-found-in-given-paths"></a>

## builds an object appending all method found in given paths


```js
const container = duckfficerMethodDi({
  Gateway: 'gateways',
  Service: 'services'
}, {
  baseDir: path.join(__dirname, './__tests__/fixtures')
})

const { OrderService } = container
t.deepEqual(await OrderService.payment({
  id: '123',
  amount: 99
}), {
  id: '123',
  result: 'paying 99$'
})
```

<a name="validates-input"></a>

## validates input


```js
const container = duckfficerMethodDi({
  Gateway: 'gateways',
  Service: 'services'
}, {
  baseDir: path.join(__dirname, './__tests__/fixtures')
})

const { PayPalGateway } = container
const { originalError } = await t.throwsAsync(() => PayPalGateway.pay('100'), {
  message: 'Invalid input'
})
t.like(originalError, {
  message: 'Invalid number'
})
```

<a name="validates-output"></a>

## validates output


```js
const container = duckfficerMethodDi({
  Gateway: 'gateways',
  Service: 'services'
}, {
  baseDir: path.join(__dirname, './__tests__/fixtures')
})

const { PayPalGateway } = container

const { originalError } = await t.throwsAsync(() => PayPalGateway.pay(0), {
  message: 'Invalid output'
})

t.like(originalError, {
  message: 'Invalid string'
})
```

<a name="emits-events"></a>

## emits events


```js
const container = duckfficerMethodDi({
  Gateway: 'gateways',
  Service: 'services'
}, {
  baseDir: path.join(__dirname, './__tests__/fixtures')
})

const { PayPalGateway } = container

PayPalGateway.on('Paid', (payload) => {
  t.log(payload)
  t.pass()
})

return PayPalGateway.pay(100)
```


<br><a name="duckfficerMethodDi"></a>

### duckfficerMethodDi(pathResolvers, [baseDir], [methodsPath]) ⇒ <code>Object</code>

| Param | Type | Default |
| --- | --- | --- |
| pathResolvers | <code>Object</code> |  | 
| [baseDir] | <code>String</code> | <code>process.cwd()</code> | 
| [methodsPath] | <code>String</code> | <code>methods</code> | 

**Returns**: <code>Object</code> - container  

* * *

### License

[MIT](https://opensource.org/licenses/MIT)

&copy; 2020-present Martin Rafael Gonzalez <tin@devtin.io>


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

### [1.0.2](https://github.com/devtin/duckfficer-method-di/compare/v1.0.1...v1.0.2) (2021-07-15)


### Chore

* **dependencies:** update ([abba97e](https://github.com/devtin/duckfficer-method-di/commit/abba97eb20112357c1b3973ed6a3d798d4921918))

### [1.0.1](https://github.com/devtin/duckfficer-method-di/compare/v1.0.0...v1.0.1) (2021-02-13)


### Bug Fixes

* remove unnecessary files ([4d88f02](https://github.com/devtin/duckfficer-method-di/commit/4d88f02da0222d8aaa62245560a0849a1c6fb887))


### Chore

* **dependencies:** update ([7a08c94](https://github.com/devtin/duckfficer-method-di/commit/7a08c94a6e99f4507b7cb34f2d4be1e26e10bb45))

## 1.0.0 (2021-02-07)


src/lib/__tests__/fixtures/gateways/pay-pal/methods/pay/input.js
module.exports = Number


src/lib/__tests__/fixtures/gateways/pay-pal/methods/pay/events.js
module.exports = {
  Paid: Object
}


src/lib/__tests__/fixtures/gateways/pay-pal/methods/pay/handler.js
module.exports = () => {
  return async (amount, { emit }) => {
    if (amount === 0) {
      return amount
    }

    await emit('Paid', {
      something: 'was paid',
      amount
    })
    return `paying ${amount}$`
  }
}


src/lib/get-local-resources.js
import path from 'path'
import startCase from 'lodash/startCase'
import { deepListDirSync } from 'deep-list-dir'
import { EventEmitter } from 'events'
import { duckfficerMethod } from 'duckfficer-method'
import { jsDirIntoJsonSync } from 'js-dir-into-json'

const upperCamelCase = (s) => {
  return startCase(s).replace(/[\s]+/g, '')
}

class DuckfficerMethodError extends Error {
  constructor (message, errors) {
    super()
    this.errors = errors
  }
}

const loadResource = ({ resourcesInstances, resourcesData, resourceName, container }) => {
  const resource = new EventEmitter()

  Object.entries(resourcesData[resourceName]).forEach(([methodName, method]) => {
    const duckfficerFn = duckfficerMethod({
      ...method,
      handler: method.handler(container)
    })
    resource[methodName] = async (payload) => {
      const result = await duckfficerFn(payload)

      result.eventsEmitted.forEach(ev => resource.emit(ev.name, ev.payload))

      if (result.errorsThrown.length > 0) {
        throw new DuckfficerMethodError(`${result.errorsThrown.length} found in ${resourceName}.${methodName}`, result.errorsThrown)
      }

      return result.output
    }
  })

  resourcesInstances[resourceName] = resource

  return resource
}

const getResolverFn = ({
  baseDir,
  resources
}) => {
  const resourcesData = {}
  const resourcesInstances = {}

  return (resourceName) => {
    if (!resources[resourceName]) {
      throw new Error(`Resource ${ resourceName } not found!`)
    }

    if (!resourcesData[resourceName]) {
      resourcesData[resourceName] = jsDirIntoJsonSync(path.join(baseDir, resources[resourceName]))
    }

    return (container) => {
      if (resourcesInstances[resourceName]) {
        return resourcesInstances[resourceName]
      }
      return loadResource({ resourcesInstances, resourcesData, resources, resourceName, container })
    }
  }
}

export const getLocalResources = (pathResolvers, {
  baseDir,
  methodsPath
}) => {
  const resolvers = {}

  Object.entries(pathResolvers).forEach(([resolverName, resolverPath]) => {
    const foundResources = {}
    const resolverPattern = path.join(resolverPath, '**', methodsPath, '**', 'handler.js')

    const handlersPaths = deepListDirSync(baseDir, {
      pattern: [resolverPattern]
    })

    handlersPaths.forEach((handlerPath) => {
      const resourceName = upperCamelCase(path.relative(baseDir, handlerPath).split('/')[1]+'-'+resolverName)
      foundResources[resourceName] = path.relative(baseDir, handlerPath).split('/').slice(0, 2).concat(methodsPath).filter(Boolean).join('/')
    })

    resolvers[resolverName] = getResolverFn({
      resolverName,
      resources: foundResources,
      baseDir
    })
  })

  return resolvers
}


src/lib/duckfficer-method-di.unit.js
import test from 'ava'
import path from 'path'
import { duckfficerMethodDi } from './duckfficer-method-di.js'

test('builds an object appending all method found in given paths', async (t) => {
  const container = duckfficerMethodDi({
    Gateway: 'gateways',
    Service: 'services'
  }, {
    baseDir: path.join(__dirname, './__tests__/fixtures')
  })

  const { OrderService } = container
  t.deepEqual(await OrderService.payment({
    id: '123',
    amount: 99
  }), {
    id: '123',
    result: 'paying 99$'
  })
})

test('validates input', async (t) => {
  const container = duckfficerMethodDi({
    Gateway: 'gateways',
    Service: 'services'
  }, {
    baseDir: path.join(__dirname, './__tests__/fixtures')
  })

  const { PayPalGateway } = container
  const { originalError } = await t.throwsAsync(() => PayPalGateway.pay('100'), {
    message: 'Invalid input'
  })
  t.like(originalError, {
    message: 'Invalid number'
  })
})

test('validates output', async (t) => {
  const container = duckfficerMethodDi({
    Gateway: 'gateways',
    Service: 'services'
  }, {
    baseDir: path.join(__dirname, './__tests__/fixtures')
  })

  const { PayPalGateway } = container

  const { originalError } = await t.throwsAsync(() => PayPalGateway.pay(0), {
    message: 'Invalid output'
  })

  t.like(originalError, {
    message: 'Invalid string'
  })
})

test('emits events', async (t) => {
  const container = duckfficerMethodDi({
    Gateway: 'gateways',
    Service: 'services'
  }, {
    baseDir: path.join(__dirname, './__tests__/fixtures')
  })

  const { PayPalGateway } = container

  PayPalGateway.on('Paid', (payload) => {
    t.log(payload)
    t.pass()
  })

  return PayPalGateway.pay(100)
})
